src/openms/include/OpenMS/DATASTRUCTURES/OSWData.h

```diff
@@ -0,0 +1,344 @@
+        {
+          return seq_;
+        }
+        /// precursor charge
+        short getCharge() const
+        {
+          return charge_;
+        }
+        /// is this a decoy feature (from a decoy protein)
+        bool isDecoy() const
+        {
+          return decoy_;
+        }
+        /// m/z of this charged peptide
+        float getPCMz() const
+        {
+          return precursor_mz_;
+        }
+        /// candidate explanations
+        const std::vector<OSWPeakGroup>& getFeatures() const
+        {
+          return features_;
+        }
+
+      private:
+        String seq_;
+        short charge_{0};
+        bool decoy_{false};
+        float precursor_mz_{0};
+        std::vector<OSWPeakGroup> features_;
+    };
+
+    /**
+      @brief A Protein is the highest entity and contains one or more peptides which were found/traced.
+
+    */
+    class OPENMS_DLLAPI OSWProtein
+    {
+      public:
+        /// just a dummy feature to allow for acceptor output values etc
+        OSWProtein() = default;
+        /// custom c'tor which fills all the members with data; all members are read-only
+        OSWProtein(const String& accession, const Size id, std::vector<OSWPeptidePrecursor>&& peptides);
+        /// Copy c'tor
+        OSWProtein(const OSWProtein& rhs) = default;
+        /// assignment operator
+        OSWProtein& operator=(const OSWProtein& rhs) = default;
+        /// move c'tor
+        OSWProtein(OSWProtein&& rhs) = default;
+        /// move assignment operator
+        OSWProtein& operator=(OSWProtein&& rhs) = default;
+
+        const String& getAccession() const
+        {
+          return accession_;
+        }
+
+        Size getID() const
+        {
+          return id_;
+        }
+
+        const std::vector<OSWPeptidePrecursor>& getPeptidePrecursors() const
+        {
+          return peptides_;
+        }
+
+      private:
+        String accession_;
+        Size id_;
+        std::vector<OSWPeptidePrecursor> peptides_;
+    };
+
+
+    /**
+      @brief Holds all or partial information from an OSW file
+
+      First, fill in all transitions and only then add proteins (which reference transitions via their transition-ids deep down).
+      References will be checked and enforced (exception otherwise -- see addProtein()).
+
+    */
+    class OPENMS_DLLAPI OSWData
+    {
+      public:
+
+        /// Adds a transition; do this before adding Proteins
+        void addTransition(const OSWTransition& tr)
+        {
+          transitions_.emplace(tr.getID(), tr);
+        }
+
+        /// Adds a protein, which has all its subcomponents already populated
+        /// All transition references internally are checked to make sure
+        /// they are valid.
+        /// You can add stub proteins, by omitting their peptide references.
+        /// @throws Exception::Precondition() if transition IDs within protein are unknown
+        void addProtein(OSWProtein&& prot);
```

Comment:
do we need a const & version? What happens again if I just pass an lvalue?

Reply:
> What happens again if I just pass an lvalue?

I does not compile and forces you to std::move.
And for now, that is exactly what I want :)

Reply:
pyOpenMS entered the conversation